keps/sig-node/4781-kublet-restart-pod-status/README.md

```diff
@@ -0,0 +1,705 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a kubelet restarts, pods with configured readiness probes may momentarily report containerNotReady status, potentially leading to service disruptions. This KEP proposes enhancing kubelet's pod readiness management to ensure that pod readiness status is preserved across kubelet restarts.
+
+## Motivation
+
+Maintaining high availability and minimizing service interruptions is crucial for Kubernetes clusters. When the kubelet restarts, it re-evaluates the state of all containers and pods. This can sometimes lead to a situation where the container's Ready state, which was previously set based on the success of its readiness probes, gets reset. This reset can cause services to be incorrectly marked as unavailable, even though they were functioning correctly before the kubelet restart.
+
+### Goals
+
+- Maintain the previous ready state of containers after kubelet restart until a new probe result is available.
+- Ensure consistency in container ready states across kubelet restarts.
+- Minimize unnecessary service disruptions caused by temporary ready state changes.
+
+### Non-Goals
+
+- Change the behavior of containers that do not have readiness probes configured.
+- Modify the fundamental logic of how readiness probes work.
+
+## Proposal
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a user of Kubernetes, I want the container's Ready state to remain consistent across kubelet restarts so that my services do not experience unnecessary downtime.
+
+### Risks and Mitigations
+
+##### Performance impact
+The additional checks and logic in the UpdatePodStatus function could potentially impact performance, especially in clusters with a large number of pods and containers.
+##### Inconsistency with other Kubernetes components
+If other parts of Kubernetes (e.g., the API server, controllers) expect certain behavior regarding container readiness states, these changes might cause inconsistencies.
+
+
+## Design Details
+
+##### Changes in kubelet_pods.go
+
+In the convertToAPIContainerStatuses function, we will preserve the previous ready state for running containers.
+```
+case cs.State == kubecontainer.ContainerStateRunning:
+    status.State.Running = &v1.ContainerStateRunning{StartedAt: metav1.NewTime(cs.StartedAt)}
+    if oldStatus != nil {
+        status.Ready = oldStatus.Ready
+    }
+```
+##### Changes in probe
+
+We will update the UpdatePodStatus function to:
```

Comment:
maybe you can also properly set `started` for a container (ref https://github.com/kubernetes/kubernetes/issues/115553)